doc/reference.md

```diff
@@ -0,0 +1,509 @@
+
+Values of different types are never considered equal – there is no implicit casting (unlike, for example, in JavaScript).
+
+We say that values `a` and `b` are *the same value*, or *equivalent* if it is impossible to tell these values apart in any way in Jsonnet. More formally `a` and `b` are equivalent if there does not exist a Jsonnet function `f` such that exactly one of `f(a)` and `f(b)` results in an error.
+
+In general, equivalent values may have different representations, and that may have performance implications, but does not affect the result.
+
+Equivalent values of course cannot be unequal, but equal values may not be equivalent (e.g. `{ a: 1, b: 1}` and `{a: 1, b: self.a}`). 
+
+### Null
+
+Null is the simplest type in Jsonnet. It has only one value –  `null`.
+
+There is no special handling of `null`, it is a value like any other.
+In particular arrays can have null elements and objects can have null fields. The null value is equal only to the null value.
+
+### Boolean 
+
+Boolean has two values `true` and `false`. They are the only values which can be used in `if` conditions.
+
+### String
+
+Strings in Jsonnet are sequences of Unicode codepoints.
+
+In most contexts they can be treated as an array of one codepoint strings (e.g. `std.length` and the `[]` operator work like that), but they are not equivalent (for example `std.type("foo") != std.type(["f", "o", "o"])`).
+
+Unlike arrays, strings are strict, meaning that evaluating a string requires calculating all its contents.
+
+Strings can be constructed as literals, slices of existing strings, concatenations of existing strings or converted from array of Unicode codepoint numbers.
+
+### Number
+
+Jsonnet numbers are 64-bit floating point numbers as defined in IEEE754 excluding `nan` and `inf` values. Operation resulting in infinity or not a number is considered an error. 
+
+Integers can be precisely represented as a Jsonnet number in the range [-2^53,2^53]. This is [a direct consequence of IEEE754 spec](https://en.wikipedia.org/wiki/Double-precision_floating-point_format).
+
+### Function
+
+Functions in Jsonnet are functions in the mathematical sense. Each function has parameters and a body expression. The result of calling a function is equivalent to the result of evaluating its body with arguments introduced to the environment.
+
+The basic way to define a Jsonnet function, is using a *function literal*:
+```
+local func = function(x) x * 2;
+func(21)
+```
+
+Jsonnet has syntax sugar for defining functions, which allows shorter, equivalent version:
+```
+local func(x) = x * 2;
+func(21)
+```
+
+The arguments are **not** evaluated when a function is called. They are passed lazily and evaluated only when actually used. This allows expressing things, which in other languages require built-in functionality or macros, such as the short-circuiting boolean operations:
```

Comment:
Partially applied. I don't want to add "in the function body", because it can be passed to another function or saved in an array without evaluating it and it can be used well after the function evaluation is finished, in the code outside of the function body.